src/tools/PICS-generator/PICSGenerator.py

```diff
@@ -0,0 +1,429 @@
+    for picsItem in serverCommandsGeneratedNode:
+        itemNumberElement = picsItem.find('itemNumber')
+
+        console.print(f"Searching for {itemNumberElement.text}")
+
+        if f"{itemNumberElement.text}" in generatedCommandPicsList:
+            console.print("Found generatedCommand PICS value in XML template ✅")
+            supportElement = picsItem.find('support')
+            supportElement.text = "true"
+
+    # Event PICS (Work in progress)
+    # The ability to set event PICS is fairly limited, due to EventList not being supported,
+    # as well as no way to check for event support in the current Matter SDK.
+
+    # This implementation marks an event as supported if:
+    # 1) Event is mandatody
+    # 2) The event is mandatory based on a feature that is supported (Cross check against feature list) (Not supported yet)
+    serverEventsNode = root.find("./clusterSide[@type='Server']/events")
+    for picsItem in serverEventsNode:
+        itemNumberElement = picsItem.find('itemNumber')
+        statusElement = picsItem.find('status')
+
+        try:
+            condition = statusElement.attrib['cond']
+            console.print(f"Checking {itemNumberElement.text} with conformance {statusElement.text} and condition {condition}")
+        except ET.ParseError:
+            condition = ""
+            console.print(f"Checking {itemNumberElement.text} with conformance {statusElement.text}")
+
+        if statusElement.text == "M":
+
+            # Is event mandated by the server
+            if condition == clusterPicsCode:
+                console.print("Found event mandated by server ✅")
+                supportElement = picsItem.find('support')
+                supportElement.text = "true"
+                continue
+
+            if condition in featurePicsList:
+                console.print("Found event mandated by feature ✅")
+                supportElement = picsItem.find('support')
+                supportElement.text = "true"
+                continue
+
+            if condition == "":
+                console.print("Event is mandated without a condition ✅")
+                continue
+
+    # Grabbing the header from the XML templates
+    inputFile = open(f"{xmlPath}{fileName}", "r")
+    outputFile = open(f"{outputPathStr}/{fileName}", "ab")
+    xmlHeader = ""
+    inputLine = inputFile.readline().lstrip()
+
+    while 'clusterPICS' not in inputLine:
+        xmlHeader += inputLine
+        inputLine = inputFile.readline().lstrip()
+
+    # Write the PICS XML header
+    outputFile.write(xmlHeader.encode())
+
+    # Write the PICS XML data
+    tree.write(outputFile)
+
+
+async def DeviceMapping(devCtrl, nodeID, outputPathStr):
+
+    # --- Device mapping --- #
+    console.print("[blue]Perform device mapping")
+    # Determine how many endpoints to map
+    # Test step 1 - Read parts list
+
+    partsListResponse = await devCtrl.ReadAttribute(nodeID, [(rootNodeEndpointID, Clusters.Descriptor.Attributes.PartsList)])
```

Comment:
For later - if you use the parsed stuff from the device basic composition here instead of reading individually, this also works on a non-commissioned device

Reply:
This tool was my first dip into scraping a device and I didn't want to rework it to do wildcard read. Are you thinking about running the test on PASE? I have noticed that if the commissioning fails it will try to do the scarping using PASE and it seemed, in the few times I ended up there, that it did work.

Reply:
Yeah - if you inherit from the BasicCompositionTests, the setup code does the wildcard read and parsing for you, so you can access all this info via the cluster object
https://github.com/project-chip/connectedhomeip/blob/master/src/python_testing/basic_composition_support.py#L99

But, as with other things, we can finesse this later. Let's just resolve this for now, and get the PR in, i think.

Reply:
Arh, I see, in that sense. Sure. We should certainly align on just using this, as you say, this sounds like a thing I can put in the backlog and have it for a 1.1 version of the tool.